Creation of a tool to delete an MRI upload and everything associated to it (#391)
aces/Loris-MRI: docs/scripts_md/delete_mri_upload.md

# NAME

delete\_mri\_upload.pl -- Delete eveything that was produced by the MRI pipeline for a given MRI upload

# SYNOPSIS

perl delete\_mri\_upload.pl \[-profile file\] \[-i\] \[-n\]

Available options are:

\-profile     : name of the config file in `../dicom-archive/.loris_mri` (defaults to `prod`).

\-i           : when performing the file backup, ignore files that do not exist or are not readable
               (default is to abort if such a file is found). This option is ignored if `-n` is used.

\-n           : do not backup the files produced by the MRI pipeline for this upload (default is to
               perform a backup).

# DESCRIPTION

This program deletes all the files and database records produced by the MRI pipeline for a given 
MRI upload. More specifically, the script will remove the records associated to the MRI upload whose
ID is passed on the command line from the following tables: `notification_spool`, `tarchive_series`
`tarchive_files`, `files_intermediary`, `parameter_file`, `files`, `mri_violated_scans`
`mri_violations_log`, `MRICandidateErrors`, `mri_upload` and `tarchive`. It will also delete from 

> the file system the files that are associated to the upload and are listed in tables `files`

Author: The script deletes the  archive listed in table `tarchive`, which is listed in the list of tables associated to the upload. Please advise on what you want me do to here.

> the file system the files that are associated to the upload and are listed in tables `files`

Author: The script deletes the records from tables `mri_violations_log`, `mri_violated_scans` and `MRICandidateErrors` but does not remove any files from `trashbin`. I'll modify the script in order to do this.

> the file system the files that are associated to the upload and are listed in tables `files`

Reviewer: > The script deletes the archive listed in table `tarchive`, which is listed in the list of tables associated to the upload. Please advise on what you want me do to here.

@nicolasbrossard it was not clear when I read the documentation that the tarchive would also be removed so I would just add that to the documentation so that it is clear that the DICOM archive is also backed up and removed from the filesystem and database

> the file system the files that are associated to the upload and are listed in tables `files`

Author: Ok, updated the script description.

`files_intermediary` and `parameter_file`. The script will abort and will not delete anything if there 

> is QC information associated to the upload (i.e entries in tables `files_qcstatus` or `feedback_mri_comments`).

Author: I assume you are talking about all the fields named `MRIQC*`. As you say, if they are set, then QC has been done on that archive and the script would not allow the deletion anyway. I am not sure I should worry about those fields. Please advise on what to do.

> is QC information associated to the upload (i.e entries in tables `files_qcstatus` or `feedback_mri_comments`).

Reviewer: I thought about this a bit more and I think it is best to not check those fields from the session table in the end so all good as it is

If the script finds a file that is listed in the database but that does not exist on the file system or is not
readable, the script will issue an error message and abort, leaving the file system and database untouched. 
This behaviour can be changed with option `-i`. By default, the script will create a backup of all the files 
that it plans to delete before actually deleting them. Use option `-n` to perform a 'hard' delete (i.e. no backup).

> The backup file name will be `mri_upload.<UPLOAD_ID`.tar.gz>. Note that the file paths inside this backup archive

Reviewer 2: ```suggestion
The backup file name will be `mri_upload.<UPLOAD_ID.tar.gz>`. Note that the file paths inside this backup archive
```

Author: ok

be stored in a `.tar.gz` archive where all paths are relative to `/` (i.e absolute paths).

INPUTS:

    - $uploadId  : ID of the upload to delete.
    - $archiveLocation: absolute path of the backed up archive created by the MRI pipeline.
    - $filesRef: reference to the array that contains all files in table C<files> associated to
                 the upload.
    - $intermediaryFilesRef: reference to the array that contains all files in table C<files_intermediary>
                             associated to the upload.
    - $picFilesRef: reference to the array that contains all files in table C<parameter_file>
                    associated to the upload.
                    

### deleteUploadInDatabase($dbh, $uploadID, $tarchiveID, $sessionID, $intermediaryFilesRef, $picFilesRef)

This method deletes all information in the database associated to the given archive. More specifically, it 
deletes records from tables `notification_spool`, `tarchive_files`, `tarchive_series`, `files_intermediary`
`parameter_file`, `files`, `mri_protocol_violated_scans`, `mri_violations_log`, `MRICandidateErrors`

> `mri_upload` and `tarchive`. It will also set the CScan\_done> value of the scan's session to 'N' if the upload

Reviewer 2: ```suggestion
`mri_upload` and `tarchive`. It will also set the C<Scan\_done> value of the scan's session to 'N' if the upload
```

Author: ok

is the last upload tied to that session. All the delete/update operations are done inside a single transaction so 
either they all succeed or they all fail (and a rollback is performed).

INPUTS:

    - $dbh       : database handle.
    - $uploadId  : ID of the upload to delete.
    - $tarchiveID: ID of the tarchive to delete.

>     - $sessionID : ID of the session associated to the scan,

Reviewer 2: ```suggestion
    - $sessionID : ID of the session associated to the scan.
```

Author: ok

    - $filesRef: reference to the array that contains all files in table C<files> associated to
                 the upload.
    - $intermediaryFilesRef: reference to the array that contains all files in table C<files_intermediary>
                             associated to the upload.
    - $picFilesRef: reference to the array that contains all files in table C<parameter_file>
                    associated to the upload.
                    

### deleteUploadFiles($archiveLocation, $filesRef, $intermediaryFilesRef, $picFilesRef)

This method deletes form the file system all the files tied to the upload that were listed in

> tables `files`, `files_intermediary` and &lt;parameter\_file>, along with the back up of the 

Reviewer 2: ```suggestion
tables `files`, `files_intermediary` and C<parameter\_file>, along with the back up of the 
```

Author: ok
